var/spack/repos/builtin/packages/petsc/package.py

```diff
@@ -589,22 +578,73 @@ def test(self):
             }
             make('ex50', parallel=False)
             for feature, featureopt in testdict.items():
-                if not feature or feature in spec:
-                    self.run_test(runexe, runopt + testexe + featureopt)
-            if '+cuda' in spec:
-                make('ex7', parallel=False)
-                testexe = ['ex7', '-mat_type', 'aijcusparse',
-                           '-sub_pc_factor_mat_solver_type', 'cusparse',
-                           '-sub_ksp_type', 'preonly', '-sub_pc_type', 'ilu',
-                           '-use_gpu_aware_mpi', '0']
-                self.run_test(runexe, runopt + testexe)
-            make('clean', parallel=False)
-        w_dir = join_path(self.install_test_root, 'src/snes/tutorials')
+                if not feature or feature in self.spec:
+                    purpose_str = 'test: run ex50 example with {0}'.format(feature)
+                    self.run_test(runexe,
+                                  options=runopt + testexe + featureopt,
+                                  purpose=purpose_str,
+                                  work_dir=w_dir)
+
+    def run_ex7_test(self, runexe, runopt, w_dir):
+        """Run stand alone test: ex7"""
+
+        make('ex7', parallel=False)
+        testexe = ['ex7', '-mat_type', 'aijcusparse',
+                   '-sub_pc_factor_mat_solver_type', 'cusparse',
+                   '-sub_ksp_type', 'preonly', '-sub_pc_type', 'ilu',
+                   '-use_gpu_aware_mpi', '0']
+        purpose_str = 'test: run ex7 example with +cuda'
+        self.run_test(runexe,
+                      options=runopt + testexe,
+                      purpose=purpose_str,
+                      work_dir=w_dir)
+        make('clean', parallel=False)
```

Comment:
The two calls to `make` here are going to be in the default test directory, not `w_dir`.  You'll want to either put this code under a `with working_dir(w_dir):` or switch to using `self.run_test()` where you can provide the `w_dir`.

Given the pattern with the other examples, I suggest using `working_dir()`.

Reply:
Since this work is now performed in the test stage, I suggest removing the `make('clean'...)` call.

Reply:
Otherwise, you'd want to move it outside `run_ex7_test` so the operation can be performed if `petsc` was not installed with `+cuda`.